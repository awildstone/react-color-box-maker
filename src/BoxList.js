import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

const BoxList = () => {
    const [ boxes, setBoxes ] = useState([]);

    const addBox = (newBox) => {
        setBoxes(boxes => [ ...boxes, newBox ]);
    };

    const removeBox = (boxId) => {
        setBoxes(boxes.filter(box => box.id !== boxId));
    }

    return (
        <div>
            <h1>Color Box Maker</h1>
            <NewBoxForm addBox={ addBox } />
            <div>
                { boxes.map(({ id, bgColor, width, height }) => 
                    <Box 
                        key={id} 
                        id={id} 
                        bgColor={bgColor} 
                        width={width}
                        height={height}
                        removeBox={removeBox}
                    />
                ) }
            </div>
        </div>
    )
}

export default BoxList;