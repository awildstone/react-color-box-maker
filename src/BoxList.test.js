import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it('renders a BoxList without crashing', () => {
    render(<BoxList />);
});

it('matches a BoxList snapshot', () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', () => {
    const { queryByText, getByLabelText } = render(<BoxList />);

    //no boxes should be in the DOM
    expect(queryByText('X')).not.toBeInTheDocument();

    const color = getByLabelText('Choose a color');
    const width = getByLabelText('Width');
    const height = getByLabelText('Height');
    const btn = queryByText('Add');

    //add the new box
    fireEvent.change(color, { target: { value: '#fff888' } });
    fireEvent.change(width, { target: { value: '200' } });
    fireEvent.change(height, { target: { value: '400' } });
    fireEvent.click(btn);

    //remove button should be present
    expect(queryByText('X')).toBeInTheDocument();

    //the box should have these styles
    expect(queryByText('X').previousSibling).toHaveStyle({backgroundColor: '#fff888', width: '200px', height: '400px'} );

    //form values should reset
    expect(color.getAttribute('value')).toBe('black');
    expect(width.getAttribute('value')).toBe('');
    expect(height.getAttribute('value')).toBe('');
});

it('can remove a box', () => {
    const { queryByText, getByLabelText } = render(<BoxList />);
    const color = getByLabelText('Choose a color');
    const width = getByLabelText('Width');
    const height = getByLabelText('Height');
    const btn = queryByText('Add');

    //add the new box
    fireEvent.change(color, { target: { value: '#fff888' } });
    fireEvent.change(width, { target: { value: '200' } });
    fireEvent.change(height, { target: { value: '400' } });
    fireEvent.click(btn);

     //remove button should be present & have correct styles
     expect(queryByText('X')).toBeInTheDocument();
     expect(queryByText('X').previousSibling).toHaveStyle({backgroundColor: '#fff888', width: '200px', height: '400px'} );

    //remove the box
    const removeBtn = queryByText('X');
    fireEvent.click(removeBtn);

    //box remove button should be gone
    expect(queryByText('X')).not.toBeInTheDocument();
});