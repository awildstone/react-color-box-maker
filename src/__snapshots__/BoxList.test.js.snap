// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a BoxList snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Color Box Maker
    </h1>
    <form>
      <label
        for="bgColor"
      >
        Choose a color
      </label>
      <input
        id="bgColor"
        name="bgColor"
        type="color"
        value="black"
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
